src/colinajose/service/SchoolService.java

```diff
@@ -61,20 +68,73 @@ public String registerTeacher(String name, String ci, int age, String gender, St
         school.addTeacher(teacher);
         return teacher.getId();
     }
-    public void sendNotifications(){}
     public void setGradingScale(String courseId, double scholarship, double expelled, double notify){
         Course course = SearchService.getCourse(this.school, courseId);
         GradesService.setGradingScale(course, scholarship, expelled, notify);
     }
-    public MyCircularDoublyLinkedList<Student> getExpelledStudents(String kardexId){
+    public void computeGrades(String kardexId){
+        Kardex kardex = SearchService.getKardex(this.school, kardexId);
+        MyCircularDoublyLinkedList<Grade> grades = GradesService.computeGrades(kardex);
+        GradesService.computeStudentStates(kardex, grades);
+    }
+    public void updateNotifyList(String kardexId){
+        MyCircularDoublyLinkedList<Device> devicesToNotify = getDevicesToNotify(kardexId);
+        registerObservers(devicesToNotify);
+        MyCircularDoublyLinkedList<Device> devicesToDetach = getDevicesToDetach(kardexId);
+        removeObservers(devicesToDetach);
+    }
+    public void sendNotifications(){
+        this.observable.notifyObservers();
+    }
+    private MyCircularDoublyLinkedList<Device> getDevicesToDetach(String kardexId){
+        MyLinkedList<Student> students = new MyLinkedList<>();
+        students.addAll(getAverageStudents(kardexId));
+        students.addAll(getScholarshipStudents(kardexId));
+        students.addAll(getExpelledStudents(kardexId));
+        return getDevices(students);
+    }
+    private MyCircularDoublyLinkedList<Device> getDevicesToNotify(String kardexId) {
+        MyLinkedList<Student> students = new MyLinkedList<>();
+        students.addAll(getNotifyStudents(kardexId));
+        return getDevices(students);
+    }
+    private MyCircularDoublyLinkedList<Device> getDevices(MyLinkedList<Student> students){
+        MyCircularDoublyLinkedList<Device> devices = new MyCircularDoublyLinkedList<>();
+        for (int i = 0; i < students.size(); i++) {
+            Student student = students.get(i);
+            MyCircularDoublyLinkedList<Person> parents = student.getParents();
+            for (int j = 0; j < parents.size(); j++) {
+                Device device = SearchService.getDevice(this.school, parents.get(i));
```

Comment:
Not sure I like this. One person can have many devices. Meaning that the owner can reference to the same parent for multiple devices.
This should be getDevices and return a Collection of devices.

Reply:
@JurgenPadilla fixed in next commit